Autocompletion

mauricel opened the pull request:
Hi,

I've been working at this a while, so I thought I should at least get what I have so far for feedback.

This pull request:
* Implements [suggest] directive for commands .
* Implements Bash and Powershell installation upon first run for current user.
* Adds necessary tests

This pull request does not:
* Implement [suggest] auto-completion for options/parameters. Still to do.

Sorry for the wait; encountered some false bugs that slowed down development, and life got busy also.

Closes #13

Tyrrrz commented:
Take your time ;)

RichiCoder1 commented:
This is looking awesome! Did have one piece of feedback though:

> Implements Bash and Powershell installation upon first run for current user.

Please don't. As a user, the only time I expect my profiles to be touched is either if A) I'm updating them or B) I'm using a package manager which has a trigger to enable completion. I would never want a tool to abitrarily touch my profiles. Providing the ability for tool makers to opt-in to having a command which will allow their end users to trigger an install or export a registration script would be much more preferable.

mauricel commented:
> This is looking awesome! Did have one piece of feedback though:
> 
> > Implements Bash and Powershell installation upon first run for current user.
> 
> Please don't. As a user, the only time I expect my profiles to be touched is either if A) I'm personally updating them or B) I'm using a package manager which has a trigger to enable completion. I would never want a tool to abitrarily touch my profiles. Providing the ability for tool makers to opt-in to having a command which will allow their end users to trigger an install or export a registration script would be much more preferable.

I had wondered what kind of reaction this would get. My thought was to let the application developer decide as follows:

```cs
new CliApplicationBuilder().AllowSuggestMode(enabled: true, autoinstall: true)
```
if enabled = false, don't install under any circumstances
if enabled = true, autoinstall = false, install is manually triggered by user as follows:

```bash
cli.exe [suggest] --install 
## alternatively, supply reference documentation 
```
if enabled = true, autoinstall = true, install upon first run

The default values would be enabled = false, autoinstall = false.

RichiCoder1 commented:
Imho, allowing auto install at all I think is a foot gun and maintenance concern. I may honestly be a bit biased however.

Tyrrrz commented:
I'm also leaning towards letting the user decide whether they want to install autocompletion or not. We can add a hint inside the help text what would give instructions on how to do it (probably just running `cli.exe [suggest]` or something). Then that command would install autosuggestions for the current shell.

mauricel commented:
> Imho, allowing auto install at all I think is a foot gun and maintenance concern. 

For libraries, I tend to err on giving developers ultimate control over their applications, however upon second thought, I think there is wisdom here. While I've tried my best to consider the edge cases for now, and while I reckon we can probably get there eventually, I don't for a second think we wouldn't be fixing bugs. 

Ok. Let's make it a user choice, at least until we've validated functionality with real users. Even then, I suspect that we will have done more than enough with any sort of install.

> I'm also leaning towards letting the user decide whether they want to install autocompletion or not.

Yup. I'm cool with that. 

> We can add a hint inside the help text what would give instructions on how to do it

I like this

> (probably just running cli.exe [suggest] or something).

Yup. We just want to watch out for the edge case where someone wants to manually write their own hooks. We wouldn't want the action that does suggestions to also modify profiles.

mauricel commented:
Updated PR as follows
- implemented parameter and option suggestions
- added documentation for [suggest] mode in Readme.md
- Added Nuget.config to work around recent CI failures
- Number of bug fixes

Known issue: Bash autocompletion does not suggest child commands 
```
Given:  a Bash shell, working in an empty directory
When: user types CliFx.Demo <tab><tab>
Then: Suggestions are: [book] instead of [book, book add, book list, book remove]
Note: cannot reproduce issue if 'book list' is renamed to 'boom' 
```
I suspect the issue is in the bash autocomplete hook, but I don't know what the issue is exactly. More investigation needed.

Tyrrrz commented:
Thanks @mauricel. I will take a closer look over the coming weeks.

Can you please write down a short but detailed reasoning behind the changes you made? I want to better understand how it works, possible edge cases, nuances you encountered and considered. Unfortunately it's hard to do by just reading the code (as it usually is, not your fault).

If you want, you can leave review comments on your code to highlight certain parts or start a discussion if you need some input. Thanks!

mauricel commented:
Thank you for considering this pull request. I've included the requested design information. You're right about the edge cases. I hope this helps., and I would appreciate any feedback/guidance you might have. 

Sorry about the large merge! Thanks!

RichiCoder1 commented:
Random aside: Probably too big for this PR (which is already big (and awesome)), but it'd be nice to eventually be able to plug into autocompletion to allow dynamic suggestions. (e.g. `git checkout f<tab>` displaying branch suggestions)

Tyrrrz commented on CliFx.Tests/SuggestDirectiveSpecs.cs:
```diff
@@ -0,0 +1,288 @@
+
+            // Act
+            var exitCode = await application.RunAsync(
+                new[] { "[suggest]", "--envvar", "CLIFX-{GUID}", "--cursor", (variableContents.Length + cursorOffset).ToString() },
```
Do we need a dynamic environment variable or can it always be the same? It's set per-process so there shouldn't be conflicts, right?

Tyrrrz commented on Readme.md:
```diff
@@ -700,6 +700,85 @@ test
+
+Once enabled, your shell must be configured to use suggest mode as follows:
+
+1. Add your application to the PATH
```
This sounds like a dealbreaker. Can we somehow work around this? As a user, I wouldn't want to add a CLI to PATH just to enable autocompletion...

Tyrrrz commented on Readme.md:
```diff
@@ -700,6 +700,85 @@ test
+Once enabled, your shell must be configured to use suggest mode as follows:
+
+1. Add your application to the PATH
+2. Add a snippet to your shell's to instruct your shell in how to generate auto-completions. This can be done as follows.
```
```suggestion
2. Install autocompletion for the current terminal:
```

Tyrrrz commented on Readme.md:
````diff
@@ -700,6 +700,85 @@ test
+2. Add a snippet to your shell's to instruct your shell in how to generate auto-completions. This can be done as follows.
+ 
+   ``` sh
+   > cmd [suggest] --install
````
Can we drop `--install` from this and make it so `cmd [suggest]` (without other arguments) performs the install?

Tyrrrz commented on CliFx.Tests/SuggestDirectiveSpecs.cs:
```diff
@@ -0,0 +1,288 @@
+                        new[] { "[suggest]", "clifx.exe", "c" }, new[] { "cmd", "cmd02" })]
+        [InlineData("supply command options if match found, regardles of other partial matches (no options defined)",
+                        new[] { "[suggest]", "clifx.exe", "cmd" }, new string[] { })]
+        public async Task Suggest_directive_suggests_commands_by_command_line_only(string usecase, string[] commandLine, string[] expected)
```
Do I understand correctly that `[suggest]` works with both arguments provided via environment variable and through command line? Do we need both?

Tyrrrz commented on CliFx.Tests/SuggestDirectiveSpecs.cs:
```diff
@@ -0,0 +1,288 @@
+            // Assert
+            exitCode.Should().Be(0);
+
+            stdOut.Split(null)
```
What does `Split(null)` do?

Tyrrrz commented on CliFx/CliFx.csproj:
```diff
@@ -34,4 +34,20 @@
   </ItemGroup>
 
+  <PropertyGroup>
+    <TargetsForTfmSpecificContentInPackage>$(TargetsForTfmSpecificContentInPackage);CopyAnalyzerToPackage</TargetsForTfmSpecificContentInPackage>
+  </PropertyGroup>
+
+  <Target Name="CopyAnalyzerToPackage">
+    <ItemGroup>
+      <TfmSpecificPackageFile Include="$(OutDir)/CliFx.Analyzers.dll" PackagePath="analyzers/dotnet/cs" BuildAction="none" />
+    </ItemGroup>
+  </Target>
```
I think this ended up being unnecessary

Tyrrrz commented on CliFx.Tests/SuggestHookInstallSpecs.cs:
```diff
@@ -0,0 +1,82 @@
+            return builder.UseConsole(FakeConsole);
+        }
+
+        [Theory]
+        [InlineData("/usr/bin/bash", ".bashrc")]
+        [InlineData("/usr/bin/pwsh", "Microsoft.PowerShell_profile.ps1")]
+        [InlineData(@"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe", "Microsoft.PowerShell_profile.ps1")]
+        public async Task Suggest_hook_is_added_when_install_parameter_is_provided(string shellPath, string expectedHookScript)
+        {
+            // Arrange
+            var fileSystem = new FakeFileSystem();
+            fileSystem.Files[shellPath] = "stub shell interpeter";
+
+            var application = TestApplicationFactory()
+                .AllowSuggestMode(true)
+                .UseFileSystem(fileSystem)
+                .Build();                          
+
+            // Act
+            var exitCode = await application.RunAsync(
+                new[] { "[suggest]", "--install" }
+            );
+
+            // Assert
+            exitCode.Should().Be(0);
+            fileSystem.FilePaths.Count(p => p.Contains(expectedHookScript))
+                      .Should().Be(1, "expect to see hook script");
+
+            var hookScript = fileSystem.Files.Where(p => p.Key.Contains(expectedHookScript)).First().Value;
+
+            hookScript.Should().Contain($"### clifx-suggest-begins-here-{application.Metadata.Title}");
+            hookScript.Should().Contain($"### clifx-suggest-ends-here-{application.Metadata.Title}");
+
```
To be honest, I don't think these tests are needed. This type of interaction is integration-heavy and trying to test it with mocks is pointless, in my opinion. Let's just test this manually and reduce the complexity (and remove `IFileSystem` as well).

Tyrrrz commented on CliFx/Suggestions/BashEnvironment.cs:
```diff
@@ -0,0 +1,55 @@
+        {
+            var safeName = new Regex("[^a-zA-Z0-9]").Replace(commandName, "");
+            return $@"
+### clifx-suggest-begins-here-{commandName}-{Version}
+# this block provides auto-complete for the {commandName} command
+# and assumes that {commandName} is on the path
+_{safeName}_complete()
+{{
+  local word=${{COMP_WORDS[COMP_CWORD]}}
+  local IFS=$'\n'
+  # generate unique environment variable
+  CLIFX_CMD_CACHE=""clifx-suggest-$(uuidgen)""
+  # replace hyphens with underscores to make it valid
+  CLIFX_CMD_CACHE=${{CLIFX_CMD_CACHE//\-/_}}
+
+  export $CLIFX_CMD_CACHE=""${{COMP_LINE}}""
+
+  local completions
+  completions=""$({commandName} ""[suggest]"" --cursor ""${{COMP_POINT}}"" --envvar $CLIFX_CMD_CACHE 2>/dev/null)""
+  if [ $? -ne 0 ]; then
+    completions=""""
+  fi
+
+  unset $CLIFX_CMD_CACHE
+
+  COMPREPLY=( $(compgen -W ""$completions"" -- ""$word"") )
+}}
+
+complete -f -F _{safeName}_complete ""{commandName}""
+
+### clifx-suggest-ends-here-{commandName}".Replace("\r", "");
```
Can we move this to an embedded .sh resource so that we can get proper highlighting? Can use something like %COMMANDNAME% etc for string templates.

Tyrrrz commented on CliFx/Suggestions/Design.md:
````diff
@@ -0,0 +1,272 @@
+```PS
+# for CliFx.Demo book add title_of_book
+
+CliFx.Demo.exe [suggest] CliFx.Demo book add title_of_book
````
Do we need to include `CliFx.Demo` after `[suggest]`? Given that it's already the application we're executing anyway.

Tyrrrz commented on CliFx/Suggestions/Design.md:
````diff
@@ -0,0 +1,272 @@
+CliFx.Demo.exe [suggest] CliFx.Demo book add title_of_book
+# completion suggestions provided, one per line, in standard output
+```
+* The command line, and the cursor position can also be supplied by environment variable for more accurate results.
````
I remember you said that the issue was that PowerShell had "breaking weirdness". Can you give some examples of when/how it happens (to explain why we use environment variable as a workaround)?

Tyrrrz commented on CliFx/Suggestions/Design.md:
````diff
@@ -0,0 +1,272 @@
+```
+The following environments are supported
+
+* Windows Powershell, detected by the presence of "C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe"
+* Ubuntu Powershell, detected by the presence of "/usr/bin/pwsh"
+* Ubuntu Bash, detected by the presence of "/usr/bin/bash"
````
If I have Ubuntu and both Powershell and Bash installed, what happens if I run `cli [suggest] --install`? Will it install on both or just Powershell?

Also, what if I'm using an entirely different terminal? I guess the command will execute successfully, but then autocomplete won't work, which may confuse the user.

Tyrrrz commented on CliFx/Suggestions/Design.md:
```diff
@@ -0,0 +1,272 @@
+
+Observations:
+1. Text passed to the `[suggest]` directive is split inconsistently when passed via the powershell hook.
+2. Cursor positioning seems difficult to predict in relation to the rest of the supplied terms. Note that CliFx is supplied with command line arguments after processing for whitespace. I couldn't find a platform indepdenant way of extracting the raw command line string. 
```
I know on .NET/Powershell `Environment.CommandLine` returns the raw string without processing.

Tyrrrz commented on CliFx/Suggestions/Design.md:
```diff
@@ -0,0 +1,272 @@
+
+# Known Issues
+
+### Bash auto-completion does not suggest child commands in empty directories
```
In empty directories?

Tyrrrz commented on NuGet.Config:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <packageSources>
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+  </packageSources>
+</configuration>
```
Can you merge `origin/master` into your branch? We added `NuGet.config` recently there.

Tyrrrz reviewed (commented):
Thanks! Finally got around to read through this. Left some comments/questions, which should be enough for first pass. Let's slowly flesh it out :)

Tyrrrz commented:
Gonna close it for now due to inactivity, but feel free to reopen if you ever feel like working on it again (and it's fine if you don't too) 🙂
